src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
 
         private BoundExpression RewriteStringConcatenationTwoExprs(SyntaxNode syntax, BoundExpression loweredLeft, BoundExpression loweredRight)
```

Comment:
>RewriteStringConcatenationTwoExprs [](http://example.com/codeflow?start=32&length=34)

For this and the other two similar helpers, I think we should make sure unit-tests cover an entire following test matrix:
- All possible combinations of input types  with all helpers available.
- Scenarios from the previous item, but with one out of the five new helpers missing (all combinations)

No need to verify IL for this test matrix, but should verify result of concatenation at runtime.

Reply:
> For this and the other two similar helpers, I think we should make sure unit-tests cover an entire following test matrix ...

I understand that creating a set of tests like that is a lot of work. The reason why I felt it is necessary, is the nature of the code changes, too many assumptions (I think I already proved many of them wrong), complicated distributed conditional logic, etc. Besides a lot of work involved, simply implementing the matrix could still miss some of the scenarios that I found problematic. It is pretty much impossible to implement test matrix that is going to provide complete coverage of all possible combinations of affected scenarios. So, I suggest to not take on this task just yet. I am planning to make some suggestions about changing the implementation. Hopefully the suggestions might result in code which is much easier to reason about and, therefore, it would be much easier get confident about overall correctness.